### ISSUE TITLE
Releasing the dispatch semaphore

### ISSUE THREAD

**Issue by bernardnormier** [icerpc/icerpc-csharp#1970]
Both IceProtocolConnection and IceRpcProtocolConnection use a semaphore (SemaphoreSlim) to implement MaxDispatches:

https://github.com/zeroc-ice/icerpc-csharp/blob/12c2f25bd3c87e203738edead72250ed38051a93/src/IceRpc/ConnectionOptions.cs#L49

We acquire this dispatch semaphore:
 - with ice, just after constructing the incoming request
 - with icerpc, before accepting a new multiplexed stream

In the normal case a bi-dir dispatch that succeeds, we release this dispatch semaphore _after_ sending the response.

I think it would make more sense to release this dispatch semaphore _before_ sending the response - once the dispatcher (user code) is complete.

My use-case is an application wtih MaxDispatches set to 1. So we allow a client to execute only 1 dispatch at a time.

Say the client performs:
```
task1 = proxy.Op1Async();
task2 = proxy.Op2Async();
await Task.WaitAll(task1, task2);
```

The total latency with the current logic is: 
  Client2ServerLatency + Op1DispatchLatency + Server2ClientLatency + Op2DispatchLatency + Server2ClientLatency
(I assume the Op1 and Op2 latency is larger than the network latency, and as a result there is no Client2Server latency for the sending of Op2)

If we release the dispatch semaphore before sending the response, the latency would be reduced to:
  Client2ServerLatency + Op1DispatchLatency + Op2DispatchLatency + Server2ClientLatency

In other words, we're not gaining anything (as far as I can tell) by preventing new dispatches while we send a response. Just adding latency.
